File: queue/CPP/queue.cpp
#include <iostream>
#define SIZE 10

class Queue{
  int a[SIZE];
  int rear;
  int front;
public:
  Queue(){
    rear = front  = -1;
  }

  void enqueue(int x);
  int dequeue();
  void display();
  ~Queue(){}
};

void Queue::enqueue(int x){
  if(front==-1) front++;
  if(rear==SIZE-1) std::cout << "Queue is full\n";
  else{
    rear++;
    a[rear] = x;
  }
}

int Queue::dequeue(){
  int i, item;
  if(rear == -1){
    std::cout << "Empty Queue";
    return 0;
  }else{
    item = a[front];
    for(i=0;i<rear; i++)a[i] = a[i+1];
  }
  rear--;
  return item;
}

/* // Second implementation of queue without shifting -> Creates holes
int Queue::dequeue(){
  int item;
  if(front == -1 || front> rear){
    std::cout << "Queue is empty\n";
  }else{
    item = a[front];
    front++;
    return item;
  }
}
*/

void Queue::display()
{
  int i;
  if(front==-1||rear==-1||front>rear)std::cout<< "Nothing to display\n";
  else{
    for(int i = front; i<= rear; i++){
      std::cout << a[i] << std::endl;
    }
  }
}

int main (int argc, char *argv[]) {
  Queue q;
  q.enqueue(5);
  q.enqueue(4);
  q.enqueue(3);
  q.enqueue(2);
  q.enqueue(1);
  q.dequeue();
  q.display();
  return 0;
}

File: queue/zig/queue.zig
const std = @import("std");

pub fn Queue(comptime T: type) type {
    return struct {
        const Self = @This();

        items: []T,
        front: usize,
        rear: usize,
        allocator: std.mem.Allocator,

        pub fn init(allocator: std.mem.Allocator, capacity: usize) !Self {
            const items = try allocator.alloc(T, capacity);
            return Self{
                .items = items,
                .front = 0,
                .rear = 0,
                .allocator = allocator,
            };
        }

        pub fn deinit(self: *Self) void {
            self.allocator.free(self.items);
        }

        pub fn enqueue(self: *Self, item: T) !void {
            if (self.isFull()) {
                return error.QueueFull;
            }
            self.items[self.rear] = item;
            self.rear += 1;
        }

        pub fn dequeue(self: *Self) !T {
            if (self.isEmpty()) {
                return error.QueueEmprt;
            }
            const item = self.items[self.front];
            self.front += 1;
            return item;
        }

        pub fn isEmpty(self: Self) bool {
            return self.front == self.rear;
        }

        pub fn isFull(self: Self) bool {
            return self.rear == self.items.len;
        }
    };
}

test "queue operations" {
    const allocator = std.testing.allocator;
    var queue = try Queue(i32).init(allocator, 3);
    defer queue.deinit();

    try queue.enqueue(1);
    try queue.enqueue(2);

    const first = try queue.dequeue();
    try std.testing.expectEqual(first, 1);
}
